### PR TITLE
[Executor] coredump: add coredump

### LINKED ISSUE #2264: Support for Wasm Coredump
## Motivation

When the Wasm instance traps, it's sometimes difficult to understand what happened. Post-mortem debugging using coredumps (which is extensively used in native environment) would be helpful for investigating and fixing crashes.

## Details

Implement Wasm coredumps as specified by https://github.com/WebAssembly/tool-conventions/blob/main/Coredump.md. Note that the spec is early and subject to changes. Feedback very welcome!

Demo from a developer's point of view: https://github.com/xtuc/wasmgdb/wiki/Demo.

## Appendix

Wasm coredump is especially useful for serverless environment where production binaries are stripped and/or have access to limited logging.

### LINKED ISSUE #3205: feat: Implementing Coredump for Post-Mortem Debugging in WebAssembly
### Summary

This project, [wasm-coredump](https://github.com/WebAssembly/tool-conventions/blob/main/Coredump.md), aims to enhance the debugging capabilities of WebAssembly by implementing a coredump feature for post-mortem analysis. When WebAssembly encounters a runtime error, it will initiate a process to collect debugging information, including stack frames, local variable values, and a snapshot of the linear memory. This data is saved in a coredump file, which, along with DWARF information, can be analyzed post-mortem to debug the error. The project involves creating or enhancing tools to generate and analyze Wasm coredumps, addressing the unique challenges of WebAssembly's memory model and execution semantics.

### Details

Expected Outcome:
- A new API and tool option to specify the saved coredump files.
- Implement the coredump inside the WasmEdge components, when a trap is triggered, there should be a whole process to log and save any information defined in the [wasm-coredump](https://github.com/WebAssembly/tool-conventions/blob/main/Coredump.md) spec.
- The coredump file must be compatible with the current toolchain, including [wasmgdb](https://github.com/xtuc/wasm-coredump/tree/main/bin/wasmgdb) and [wasm-edit](https://github.com/xtuc/wasm-coredump/tree/main/bin/rewriter).
- Documentation on the coredump format, including how to encode process, module, and instance information.
- Tutorials, examples, and demonstration of post-mortem debugging using the implemented coredump feature.

Recommended Skills:
- WebAssembly
- C++
- Debugger

### Appendix

The original issue is from here: https://github.com/WasmEdge/WasmEdge/issues/2264

### PR DESCRIPTION
**Pull request by grorge123**
Fixes #2264 
Fixes #3205 

Support the core dump when WebAssembly enters a trap.

This core dump feature follows the [conventions](https://github.com/WebAssembly/tool-conventions/blob/main/Coredump.md). And add a parameter to generate the core dump file for the [Wasmgdb](https://github.com/xtuc/wasm-coredump).
The current implementations support most sections and use the Wasmgdb test. Limited by the architecture, module information only shows the current module.

### CODE REVIEW

**Comment by juntao** [comment]
Hello, I am a [code review agent](https://github.com/flows-network/github-pr-review/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

**Comment by hydai** [comment]
@q82419 This is ready to be reviewed. PTAL.

File: include/executor/coredump.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+void generateCoredump(const Runtime::StackManager &StackMgr,
+                      bool ForWasmgdb) noexcept;
+AST::CustomSection createCore();
+AST::CustomSection createCoremodules(
+    Span<const Runtime::Instance::ModuleInstance *const> ModuleInstances);
+AST::CustomSection
+createCorestack(Span<const Runtime::StackManager::Frame> Frames,
+                Span<const Runtime::StackManager::Value> ValueStack,
+                bool ForWasmgdb);
+AST::GlobalSection createGlobals(
```
Comment (q82419) [inline include/executor/coredump.h:20]: Sorting these functions with the instance order. (i.e. Func, Table, Memory, Global ..)

File: lib/executor/coredump.cpp [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+#include "common/errcode.h"
+#include "common/types.h"
+#include "loader/serialize.h"
+#include "runtime/stackmgr.h"
+#include "spdlog/spdlog.h"
+#include <cstdint>
+#include <cstring>
+#include <vector>
+
+#define ENCODE_LEB128(Content, Value)                                          \
```
Comment (q82419) [inline lib/executor/coredump.cpp:12]: You included the `serialize.h`.
Why not just use the `serializeUN()` or `serializeU32()`?

**Review by q82419 — CHANGES_REQUESTED** [review]

File: lib/executor/coredump.cpp (hunk 2 above)
Comment (q82419) [reply lib/executor/coredump.cpp:12]: Yes, you can export them from the serializer.

**Review by q82419 — DISMISSED** [review]

**Review by hydai — CHANGES_REQUESTED** [review]
@grorge123 
Please fix the windows issues.

**Review by hydai — APPROVED** [review]
LGTM

**Pull request merged**